cloudinit/sources/DataSourceLXD.py

```diff
@@ -223,14 +224,129 @@ def is_platform_viable() -> bool:
+        # cloud-init.* keys will sort before user.* keys which is preferred
+        # precedence.
+        for config_route in sorted(config_routes):
+            config_route_url = url_helper.combine_url(LXD_URL, config_route)
+            config_route_response = _do_request(
+                session, config_route_url, do_raise=False
+            )
+
+            if not config_route_response.ok:
+                LOG.debug(
+                    "Skipping %s on [HTTP:%d]:%s",
+                    config_route_url,
+                    config_route_response.status_code,
+                    config_route_response.text,
+                )
+                continue
+
+            cfg_key = config_route.rpartition("/")[-1]
+            # Leave raw data values/format unchanged to represent it in
+            # instance-data.json for cloud-init query or jinja template
+            # use.
+            config["config"][cfg_key] = config_route_response.text
+            # Promote common CONFIG_KEY_ALIASES to top-level keys.
+            if cfg_key in CONFIG_KEY_ALIASES:
+                # Due to sort of config_routes, promote cloud-init.*
+                # aliases before user.*. This allows user.* keys to act as
+                # fallback config on old LXD, with new cloud-init images.
+                if CONFIG_KEY_ALIASES[cfg_key] not in config:
+                    config[
+                        CONFIG_KEY_ALIASES[cfg_key]
+                    ] = config_route_response.text
+                else:
+                    LOG.warning(
+                        "Ignoring LXD config %s in favor of %s value.",
+                        cfg_key,
+                        cfg_key.replace("user", "cloud-init", 1),
+                    )
+        return config
+
+    def _get_devices(self, session: requests.Session) -> dict:
+        url = url_helper.combine_url(self._version_url, "devices")
+        response = _do_request(session, url)
+        try:
+            devices = response.json()
+        except JSONDecodeError as exc:
+            raise sources.InvalidMetaDataException(
+                "Unable to determine LXD devices from {route}."
+                " Expected JSON but found: {resp}".format(
+                    route=url, resp=response.text
+                )
+            ) from exc
+        return {"devices": devices}
+
+    def __call__(
+        self, *, metadata_only: bool = False, devices: bool = False
+    ) -> dict:
+        with requests.Session() as session:
+            session.mount(self._version_url, LXDSocketAdapter())
+            md: dict = {}
+            md.update(self._get_meta_data(session))
+            if metadata_only:
+                return md  # Skip network-data, vendor-data, user-data
+            # Document API version read
+            md["_metadata_api_version"] = self.api_version
+            md.update(self._get_config(session))
+            if devices:
+                md.update(self._get_devices(session))
```

Comment:
For the hotplug case where we potentially trigger python invocation from a udev event, we'd want this operation to only touch the devices route I presume to avoid two hits on the metadata service for every udev event. I wonder if we might want to consolidate `metadata_only` and `devices` bools into a single parameter `fitler_keys: List[str]` or `request_keys` or `filter_routes` or some other parameter name which would limit the API routes contacted to just the desired keys. Then we have flexibility to get all, some or only one API route depending on our needs

Reply:
Good point. I have implemented an `enum.Flag` to control which keys to retrieve.